src/node.ts

```diff
@@ -100,27 +118,52 @@ export function register(esbuildOptions: RegisterOptions = {}) {
     const options = getOptions(dir)
     format = format ?? inferPackageFormat(dir, filename)
 
-    const {
-      code: js,
-      warnings,
-      map: jsSourceMap,
-    } = transformSync(code, {
-      sourcefile: filename,
-      sourcemap: 'both',
-      loader: getLoader(filename),
-      target: options.target,
-      jsxFactory: options.jsxFactory,
-      jsxFragment: options.jsxFragment,
-      format,
-      ...overrides,
-    })
-    map[filename] = jsSourceMap
-    if (warnings && warnings.length > 0) {
-      for (const warning of warnings) {
-        console.log(warning.location)
-        console.log(warning.text)
+    let js = ''
+    let warnings: Message[]
+    if (!bundle) {
+      const result = transformSync(code, {
+        sourcefile: filename,
+        sourcemap: 'both',
+        loader: getLoader(filename),
+        target: options.target,
+        jsxFactory: options.jsxFactory,
+        jsxFragment: options.jsxFragment,
+        format,
+        ...overrides,
+      })
+      js = result.code
+      map[filename] = result.map
+      warnings = result.warnings
+    } else {
+      const result = buildSync({
+        entryPoints: [filename],
+        bundle: true,
+        sourcemap: 'inline',
+        platform: 'node',
+        write: false,
+        target: options.target,
+        jsxFactory: options.jsxFactory,
+        jsxFragment: options.jsxFragment,
+        format,
+        ...(typeof bundle === 'object' ? bundle : null),
```

Comment:
maybe we should mark `node_modules` as external

Reply:
we can also polyfill __dirname like this: https://github.com/egoist/bundle-require/blob/main/src/index.ts#L132

Reply:
Plugins are not supported in `buildSync()`.

Reply:
hmm it should be supported by `buildSync`, it's not supported by `transform` and `transformSync`

Reply:
https://github.com/evanw/esbuild/blob/03051a8ff4a7efccf31a9e076a46130c7dde4b20/lib/shared/common.ts#L976

Reply:
I have an idea to infer the `external` options of esbuild to try its best to exclude node_modules.

Maybe we have to lookup several times to make it work in monorepo.

Reply:
we can still support `__dirname` by using a simple string.replace 🤔